File: solutions/s2/python/pqueue.py
"""A persistent FIFO queue built from two cons-lists.

Cons-lists are exactly the ones from the lecture:
    None                 the empty list
    (head, tail)         a node whose tail is another cons-list

A queue is a pair (front, back):
    - dequeue takes from the head of `front`
    - enqueue pushes onto the head of `back`
    - abstraction:  queue as a sequence  =  front ++ reverse(back)
    - invariant:    if front is empty, back is empty too
                    (so peek/dequeue only ever look at front)

All operations return NEW queues; no argument is ever modified.
"""

EMPTY = (None, None)


def is_empty(q):
    front, _back = q
    return front is None


def peek(q):
    """The element that dequeue would return, without removing it."""
    front, _back = q
    if front is None:
        raise IndexError("peek at empty queue")
    return front[0]


def _rev(lst):
    """Reverse a cons-list, returning a new cons-list.

        _rev((1, (2, (3, None))))  ==  (3, (2, (1, None)))

    An imperative loop inside a persistent data structure looks like a
    contradiction. It is not, and the reason is a distinction worth saying
    out loud, because the rest of the course leans on it:

      rebinding a NAME      out = (head, out)
          makes `out` point at a new cell. Whatever it pointed at before is
          still there, unchanged, and anyone else holding it is unaffected.

      mutating a VALUE      cell[1] = something
          writes into a cell that other versions may be sharing. THIS is
          the one that breaks safe sharing — and tuples make it impossible,
          which is exactly why the lecture built cons cells out of them.

    Only the first kind happens here. Every `(head, out)` is a fresh cell,
    written once when it is constructed and never again.

    So: which of the five routes on the "immutability is one route, not the
    only one" slide is this? None of them. Those routes all answer the
    question "the structure IS mutated — why is sharing still safe?", and
    nothing here is mutated. A loop over local names is not a mutation of
    data at all. (OCaml makes the difference syntactic: rebinding is `let`,
    while writing a value needs `mutable` or `ref`. Which is why the OCaml
    solution to this same function is `List.rev` and raises no question.)

    The nearest row is OWNERSHIP, and it becomes the right answer the moment
    this is written with mutable nodes instead — a class with an assignable
    `tail`, the way one would in C or Java:

        node.tail = out          # a real write, to a real cell

    That is safe, and for a reason a compiler can state: the cells are
    freshly allocated, nobody else holds a pointer to them yet, and they
    become shared only at `return`. Mutate while owned, publish as
    immutable. Rust's &mut-then-freeze, Haskell's runST, Clojure's
    transients, all the same shape — and how a production library would
    write this.

    It is NOT the BENIGN EFFECTS row, which is the much harder claim: the
    cell is already shared, we write to it anyway, and we argue that no
    client can tell. That row comes due twice later — session 5's thunk
    that overwrites itself once forced (the repair for Part D), and session
    7's path compression in a persistent Union-Find. Those need an argument.
    This needs only the observation that a tuple cannot be written to.

    For a case of genuine, owned mutation in this very file, see to_list.
    """
    out = None
    while lst is not None:
        head, tail = lst
        out = (head, out)
        lst = tail
    return out


def _make(front, back):
    """Smart constructor: restores the invariant 'front empty => back empty'.

    Every queue in the program is built through this function, so the
    invariant holds everywhere. This is where the O(n) reversal hides.
    """
    if front is None:
        return (_rev(back), None)
    return (front, back)


def enqueue(q, x):
    """Return a new queue with x added at the back. O(1)."""
    front, back = q
    return _make(front, (x, back))


def dequeue(q):
    """Return (value, new_queue). Amortized O(1) — see the handout.

    Raises IndexError on the empty queue.
    """
    front, back = q
    if front is None:
        raise IndexError("dequeue from empty queue")
    head, front_rest = front
    return head, _make(front_rest, back)


# ---------------------------------------------------------------- helpers

def to_list(q):
    """The abstraction function: the sequence this queue represents.

    The tests compare the implementation against this function, which makes
    it the specification — so it is deliberately written the dumb way. Walk
    `front` collecting elements, walk `back` collecting elements, reverse
    the second and append. It mirrors the abstraction line at the top of the
    file, front ++ reverse(back), closely enough to be checked against it by
    eye. Nothing here is clever, and nothing here should become clever: a
    bug in this function would be invisible in precisely the cases it exists
    to catch.

    It is blind to the invariant, on purpose. Two different representations
    can denote the same sequence — (front=[1], back=[3,2]) and
    (front=[1,2,3], back=[]) both mean 1, 2, 3 — and to_list sends both to
    the same list. That is what makes it an abstraction function rather than
    an equality test, and it is why the persistence tests need `is` on top:
    to_list can see that the contents are right, but it cannot see whether
    the cells were shared or rebuilt.

    One thing to notice, against _rev above. This function really does
    mutate: `out.append(...)` writes into a Python list, a mutable object,
    over and over. That is not name rebinding — it is the real thing. It is
    safe by OWNERSHIP, the third row of the slide: `out` is created here,
    no other code has a reference to it, and it does not escape until
    `return`. The cons cells being read are never touched. So the file
    contains both cases side by side — no mutation at all in _rev, owned
    mutation here — and neither one needs immutability to be safe.
    """
    front, back = q
    out = []
    while front is not None:
        out.append(front[0])
        front = front[1]
    tail = []
    while back is not None:
        tail.append(back[0])
        back = back[1]
    return out + tail[::-1]


def from_iter(xs):
    q = EMPTY
    for x in xs:
        q = enqueue(q, x)
    return q

File: solutions/s2/java/PQueue.java
import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;

/**
 * SOLUTION — persistent FIFO queue over two cons-lists. Do not ship.
 *
 * Note how Java enforces the discipline that Python could only ask for
 * politely: every field is {@code final} (no accidental writes), the node
 * type is a {@code record} (immutable by construction), and the class is
 * {@code final} (no subclass can smuggle mutability back in) — compare
 * with the lecture slide "why private and final?".
 *
 * representation:  (front, back), two immutable singly-linked lists
 * abstraction:     the sequence  front ++ reverse(back)
 * invariant:       if front is null (empty), back is null too
 *
     * The four holes of the student file, filled in. To check:
 *     javac *.java && java Tests
 */
public final class PQueue<T> {

    /** A cons cell: immutable, with an automatically final head and tail. */
    private record Node<T>(T head, Node<T> tail) {}

    /** What dequeue returns: the value AND the new version of the queue. */
    public record Dequeued<T>(T value, PQueue<T> rest) {}

    private final Node<T> front;
    private final Node<T> back;

    private PQueue(Node<T> front, Node<T> back) {
        this.front = front;
        this.back = back;
    }

    private static final PQueue<?> EMPTY = new PQueue<>(null, null);

    /** The one shared empty queue — safe to share precisely because no
     *  operation can ever modify it. (Compare: the imperative interface
     *  needs a fresh {@code new} per queue.) */
    @SuppressWarnings("unchecked")
    public static <T> PQueue<T> empty() {
        return (PQueue<T>) EMPTY;
    }

    public boolean isEmpty() {
        return front == null;
    }

    /** The element dequeue would return, without removing it. */
    public T peek() {
        if (front == null) throw new NoSuchElementException("peek at empty queue");
        return front.head();
    }

    /** SOLUTION 1 — reverse a cons-list, returning a new cons-list.
     *  A plain loop over local variables is fine, and for the same reason
     *  it is fine in the Python version: only the local names move. Every
     *  {@code new Node<>} is a fresh cell, and a record has no setters, so
     *  no existing cell is ever written to. See the long note on
     *  {@code _rev} in the Python solution for which row of the
     *  "immutability is one route, not the only one" slide this is — the
     *  answer is none of them, and why is the interesting part. */
    private static <T> Node<T> rev(Node<T> lst) {
        Node<T> out = null;
        while (lst != null) {
            out = new Node<>(lst.head(), out);
            lst = lst.tail();
        }
        return out;
    }

    /** SOLUTION 2 — smart constructor: restores the invariant, that an
     *  empty front forces an empty back. Every queue is built through it,
     *  so the invariant holds everywhere; {@code private static} is what
     *  keeps clients from building one that violates it. This is where the
     *  O(n) reversal hides. */
    private static <T> PQueue<T> make(Node<T> front, Node<T> back) {
        if (front == null) return new PQueue<>(rev(back), null);
        return new PQueue<>(front, back);
    }

    /** SOLUTION 3 — a new queue with x at the back, O(1).
     *  One cell allocated; {@code back} becomes its tail and {@code front}
     *  is passed through untouched. Both halves of the old version are
     *  shared, not copied — the lecture's sharing diagram, in one line. */
    public PQueue<T> enqueue(T x) {
        return make(front, new Node<>(x, back));
    }

    /** SOLUTION 4 — the front element AND the new version, as a pair.
     *  Java has no option type here, so emptiness is an exception rather
     *  than something the compiler makes the caller face; compare the
     *  OCaml signature, {@code 'a t -> ('a * 'a t) option}. Returning
     *  {@code Dequeued} rather than mutating is the whole point: the queue
     *  we were handed is still valid afterwards. */
    public Dequeued<T> dequeue() {
        if (front == null)
            throw new NoSuchElementException("dequeue from empty queue");
        return new Dequeued<>(front.head(), make(front.tail(), back));
    }

    // ------------------------------------------------------------ helpers

    /** The abstraction function: the sequence this queue represents.
     *  The tests compare the implementation against it, so it stays
     *  obvious rather than efficient. Note that it is blind to the
     *  invariant — two different representations with the same contents
     *  give the same list — which is why persistence is tested separately,
     *  by reference identity. Unlike {@code rev}, this one genuinely
     *  mutates: {@code out.add(...)} writes into an ArrayList. That is safe
     *  by OWNERSHIP — the list is created here and does not escape until
     *  the return. */
    public List<T> toList() {
        List<T> out = new ArrayList<>();
        for (Node<T> n = front; n != null; n = n.tail()) out.add(n.head());
        List<T> tail = new ArrayList<>();
        for (Node<T> n = back; n != null; n = n.tail()) tail.add(n.head());
        for (int i = tail.size() - 1; i >= 0; i--) out.add(tail.get(i));
        return out;
    }

    @SafeVarargs
    public static <T> PQueue<T> of(T... xs) {
        PQueue<T> q = empty();
        for (T x : xs) q = q.enqueue(x);
        return q;
    }
}
